src/libstrongswan/plugins/openssl/openssl_kem.c

```diff
@@ -0,0 +1,360 @@
+		return TRUE;
+	}
+	return FALSE;
+}
+
+static bool openssl_kem_decapsulate(private_openssl_kem_t *this,
+																		chunk_t ciphertext)
+{
+	EVP_PKEY_CTX *ctx = NULL;
+	size_t shared_secret_length = 0;
+	ctx = EVP_PKEY_CTX_new(this->pkey, NULL);
+	if (!ctx)
+	{
+		return FALSE;
+	}
+	EVP_PKEY_decapsulate(ctx, NULL, &shared_secret_length, ciphertext.ptr,
+											 ciphertext.len);
+	this->shared_secret = chunk_alloc(shared_secret_length);
+	if (!EVP_PKEY_decapsulate(ctx, this->shared_secret.ptr,
+														&this->shared_secret.len, ciphertext.ptr,
+														ciphertext.len))
+	{
+		EVP_PKEY_CTX_free(ctx);
+		chunk_clear(&this->shared_secret);
+		return FALSE;
+	}
+	EVP_PKEY_CTX_free(ctx);
+	return TRUE;
+}
+
+static bool openssl_kem_encapsulate(private_openssl_kem_t *this,
+																		chunk_t public_key)
+{
+	EVP_PKEY_CTX *ctx = NULL;
+	EVP_PKEY *pkey = NULL;
+	size_t shared_secret_length = 0;
+	size_t ciphertext_length = 0;
+	size_t seed_length = 0;
+	chunk_t seed = chunk_empty;
+	drbg_t *drbg = this->thread_drbg->get(this->thread_drbg);
+
+	pkey = EVP_PKEY_kem_new_raw_public_key(openssl_kem_get_nid(this),
+																				 public_key.ptr, public_key.len);
+	if (!pkey)
+	{
+		return FALSE;
+	}
+	ctx = EVP_PKEY_CTX_new(pkey, NULL);
+	if (!ctx)
+	{
+		EVP_PKEY_free(pkey);
+		return FALSE;
+	}
+	if (drbg)
+	{
+		EVP_PKEY_encapsulate_deterministic(ctx, NULL, &ciphertext_length, NULL,
+																			 &shared_secret_length, NULL,
+																			 &seed_length);
+		this->shared_secret = chunk_alloc(shared_secret_length);
+		this->ciphertext = chunk_alloc(ciphertext_length);
+		seed = chunk_alloc(seed_length);
+		drbg->generate(drbg, seed.len, seed.ptr);
+		if (!EVP_PKEY_encapsulate_deterministic(ctx, this->ciphertext.ptr,
+																						&this->ciphertext.len,
+																						this->shared_secret.ptr,
+																						&this->shared_secret.len, seed.ptr,
+																						&seed.len))
+		{
+			EVP_PKEY_free(pkey);
+			EVP_PKEY_CTX_free(ctx);
+			chunk_clear(&this->shared_secret);
+			chunk_clear(&this->ciphertext);
+			chunk_clear(&seed);
+			return FALSE;
+		}
+	}
+	else
+	{
+		EVP_PKEY_encapsulate(ctx, NULL, &ciphertext_length, NULL,
+												 &shared_secret_length);
+		this->shared_secret = chunk_alloc(shared_secret_length);
+		this->ciphertext = chunk_alloc(ciphertext_length);
+		if (!EVP_PKEY_encapsulate(ctx, this->ciphertext.ptr, &this->ciphertext.len,
+															this->shared_secret.ptr,
+															&this->shared_secret.len))
+		{
+			EVP_PKEY_free(pkey);
+			EVP_PKEY_CTX_free(ctx);
+			chunk_clear(&this->shared_secret);
+			chunk_clear(&this->ciphertext);
+			return FALSE;
+		}
+	}
+	EVP_PKEY_free(pkey);
+	EVP_PKEY_CTX_free(ctx);
+	chunk_clear(&seed);
+	return TRUE;
+}
+
+METHOD(key_exchange_t, set_public_key, bool,
+			 private_openssl_kem_t *this, chunk_t value)
+{
+
+	if (this->shared_secret.ptr != NULL || this->ciphertext.ptr != NULL)
```

Comment:
While our code calls `set_public_key()` only once, this seems like a good call from an API perspective. I might add that to the other KEM implementations as well.